modules/s3-articfact/main.tf

```diff
@@ -0,0 +1,380 @@
+resource "aws_s3_bucket_public_access_block" "primary" {
+  bucket = aws_s3_bucket.primary.id
+
+  block_public_acls       = true
+  block_public_policy     = true
+  ignore_public_acls      = true
+  restrict_public_buckets = true
+}
+
+# -------------------------------------------
+# CONFIGURE PRIMARY S3 BUCKET SERVER SIDE ENCRYPTION
+# -------------------------------------------
+
+resource "aws_s3_bucket_server_side_encryption_configuration" "primary" {
+  bucket = aws_s3_bucket.primary.id
+
+  rule {
+    apply_server_side_encryption_by_default {
+      sse_algorithm = "aws:kms"
+    }
+  }
+}
+
+
+# ------------------------------------------------------------
+
+# THE FOLLOWING SECTION IS USED TO CREATE THE REPLICA S3 BUCKET
+
+# AND ASSOCIATED RESOURCES (LIFECYCLE MANAGEMENT, VERSIONING).
+
+# ------------------------------------------------------------
+
+# -------------------------------------------
+# CREATE S3 REPLICATION BUCKET
+# -------------------------------------------
+
+resource "aws_s3_bucket" "replica" {
+  provider = aws.replica
+  bucket   = var.replica_bucket_name
+}
+
+# -------------------------------------------
+# CONFIGURE REPLICA S3 BUCKET POLICY
+# -------------------------------------------
+
+resource "aws_s3_bucket_acl" "replica" {
+  provider = aws.replica
+  bucket   = aws_s3_bucket.replica.id
+  acl      = var.replica_bucket_acl
+}
+
+# -------------------------------------------
+# ENABLE REPLICA OBJECT VERSIONING
+# -------------------------------------------
+
+resource "aws_s3_bucket_versioning" "replica" {
+  provider = aws.replica
+  bucket   = aws_s3_bucket.replica.id
+  versioning_configuration {
+    status = var.bucket_versioning_configuration_status
+  }
+}
+
+# -------------------------------------------
+# CONFIGURE REPLICA S3 BUCKET PUBLIC ACCESS
+# -------------------------------------------
+
+resource "aws_s3_bucket_public_access_block" "replica" {
+  bucket = aws_s3_bucket.replica.id
+
+  block_public_acls       = true
+  block_public_policy     = true
+  ignore_public_acls      = true
+  restrict_public_buckets = true
+}
+
+# -------------------------------------------
+# CONFIGURE REPLICA S3 BUCKET SERVER SIDE ENCRYPTION
+# -------------------------------------------
+
+resource "aws_s3_bucket_server_side_encryption_configuration" "replica" {
+  provider = aws.replica
+  bucket   = aws_s3_bucket.replica.id
+
+  rule {
+    apply_server_side_encryption_by_default {
+      sse_algorithm = "aws:kms"
+    }
+  }
+}
+
+# -------------------------------------------
+# CREATE BUCKET REPLICATION CONFIGURATION
+# -------------------------------------------
+
+resource "aws_s3_bucket_replication_configuration" "replica" {
+  provider = aws.replica
+
+  depends_on = [aws_s3_bucket_versioning.source]
+
+  role   = aws_iam_role.replica.arn
+  bucket = aws_s3_bucket.primary.id
+
+  rule {
+    id = var.bucket_replication_configuration_rule_id
+
+    status = var.replica_configuration_status
```

Comment:
Can we change this to enable_replica. Expanding on the comment above. We shouldn't create the resource at all unless count = var.enable_replica ? 1 : 0. Therefore with that consideration we can hardcode the status.